core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -574,7 +576,22 @@ private void writeToSink(TaskId taskId, List<Slice> pages)
             verify(writeBuffer != null, "writeBuffer is expected to be initialized");
             for (Slice page : pages) {
                 // wait for the sink to unblock
-                getUnchecked(exchangeSink.isBlocked());
+                while (true) {
+                    try {
+                        exchangeSink.isBlocked().get(1, SECONDS);
+                        break;
+                    }
+                    catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
+                    }
+                    catch (ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
+                    catch (TimeoutException e) {
+                        updateSinkInstanceHandleIfNecessary();
+                    }
```

Comment:
Can you explain the logic here?

Reply:
When an `ExchangeSink` is blocked there's a change that the handle needs to be updated. This waits for some amount of time, and if the `ExchangeSink` is not unblocked - we check if the handle needs to get updated.

Reply:
When we refer to a handle update, it means that we switch to a different buffer node to write to? But we only switch to a different buffer node when the old buffer node is offline or is being drained, right? How does it related to how much time we are blocked?

Reply:
It doesn't. The idea is that when an `ExchangeSink` needs a handle update it will get blocked as it will no longer able to accept new data. Do you see any use case when we would like to refresh the handle while the `ExchangeSink` is still able to accept some data?